### ISSUE TITLE
Salvage parts of GranMinigun's unfinished SDL & shader related rework

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#3971]
As per title. It doesn't seem likely he'll come back anytime soon and pick these up again (judging by his public GitHub activity, at least), and there are some good improvements there. I'm intending to do this at some point, but if anyone else feels the urge, be my guest 😎 

(But cc'ing you anyway, man, just in case 😄  @GranMinigun)

Needs careful manual application of the changes with some slight further rework, perhaps.

These are the commit in question:

- https://github.com/dosbox-staging/dosbox-staging/pull/2450/commits/30d31f243be2a886388812bf5d49cd5a52f37321
- https://github.com/dosbox-staging/dosbox-staging/pull/2450/commits/c9f5de28b4a1f34c97fecd34bb4fe01f10b05616
- https://github.com/dosbox-staging/dosbox-staging/pull/2997/commits/fe713c0da8882b86dd3e60e77a946130f99b0693

Make sure to read the PR descriptions too for context (e.g., dropping the ancient fixed pipeline OpenGL stuff is definitely a good idea and should happen):

- https://github.com/dosbox-staging/dosbox-staging/pull/2997
- https://github.com/dosbox-staging/dosbox-staging/pull/2450

**Comment by Grounded0** [comment]
![image](https://github.com/user-attachments/assets/3ffcc57f-038e-49dd-94dc-8ab8deff1b29)

**Comment by Grounded0** [comment]
Sir GranMinigun has been notified on IRC.

**Comment by weirddan455** [comment]
![sdlmain](https://github.com/user-attachments/assets/c4acf1cb-f247-4e0b-8608-d7babf28272c)

**Comment by weirddan455** [comment]
On a more serious note, https://github.com/dosbox-staging/dosbox-staging/commit/30d31f243be2a886388812bf5d49cd5a52f37321 is something I noticed myself and mentioned on Discord.  I agree with removing those fixed function OpenGL calls.  I wasn't aware he had already started work on doing so.  I think that code rarely, if ever, gets run on any system.  It's used as a fallback in case the shaders fail to compile.  That should only happen if we have a bug in the shader code (if so, we fix the bug) or the user has a problem with their OpenGL drivers (if so, `output = texture` is a better fallback).

Regarding the shader loading changes, didn't you mention to me @johnnovak that you had a plan to merge the shaders into one mega shader or something?  I just don't want people duplicating work if you're planning on implementing something different than what's in these commits.

As for the rest, it looks like a lot of large refactoring commits.  Probably some good work in there but personally I don't really like doing refactors without some specific improvements in mind.  If someone else wants to, more power to them.  That code could definitely use some improved readability.  Just keep in mind it'll need to be refactored again if and when we migrate to SDL 3.  Maybe some cleanup before that would help the transition though.

**Comment by johnnovak** [comment]
> On a more serious note, [30d31f2](https://github.com/dosbox-staging/dosbox-staging/commit/30d31f243be2a886388812bf5d49cd5a52f37321) is something I noticed myself and mentioned on Discord. I agree with removing those fixed function OpenGL calls. I wasn't aware he had already started work on doing so. I think that code rarely, if ever, gets run on any system. It's used as a fallback in case the shaders fail to compile. That should only happen if we have a bug in the shader code (if so, we fix the bug) or the user has a problem with their OpenGL drivers (if so, `output = texture` is a better fallback).

Absolutely! 💯 

> Regarding the shader loading changes, didn't you mention to me @johnnovak that you had a plan to merge the shaders into one mega shader or something? I just don't want people duplicating work if you're planning on implementing something different than what's in these commits.

That's true, but the caching would be still useful. There will be still a few shaders around after the merge of the adaptive CRT shader variants into a single shader, e.g., `sharp`, `nearest-neighbour`, etc. will always be separate. Plus the commit also contains other good cleanup enhancements; caching the compiled shaders is a small part of it.

> As for the rest, it looks like a lot of large refactoring commits. Probably some good work in there but personally I don't really like doing refactors without some specific improvements in mind. If someone else wants to, more power to them. That code could definitely use some improved readability. Just keep in mind it'll need to be refactored again if and when we migrate to SDL 3. Maybe some cleanup before that would help the transition though.

Yeah, and I don't disagree with you. I don't want to just massage the current `sdlmain.cpp` for nothing, but if some of those refactorings are sane and improve clarity, that would make the SDL3 transition easier. In any case, I care about the refactoring commits the least, so we're on the same page.

**Comment by weirddan455** [comment]
>  That's true, but the caching would be still useful.

OK, I think I see what it's doing now.  It definitely seems like a good idea to keep the compiled shaders around rather than re-compiling on every mode change.  They should be kept in VRAM ready to go with that improvement.  We only have 37 shader files (I just counted) so that's small potatoes in terms of RAM/VRAM usage.

>  All in all, this ticket is all about selectively getting in the objectively desirable improvements GranMinigun did in those commits which we would do anyway sooner or later, so I thought, well, why not take advantage of his efforts then.

Agreed.  Good to raise awareness of this.  I may take a shot at removing those fixed function OpenGL calls.  It looks fairly straight-forward to do.  After 0.82.0 though.  I don't want to touch SDL pain this close to release :laughing:

**Comment by Grounded0** [comment]
> Yeah, and I don't disagree with you. I don't want to just massage the current sdlmain.cpp for nothing, but if some of those refactorings are sane and improve clarity, that would make the SDL3 transition easier. In any case, I care about the refactoring commits the least, so we're on the same page.

Hey that PR provides a nice 4 percent speed bump so that's nice.

**Comment by johnnovak** [comment]
> I may take a shot at removing those fixed function OpenGL calls. It looks fairly straight-forward to do.

Go for it, man! Eliminate those fixed-pipeline calls with _extreme prejudice!_ 😎 💪🏻 Maybe another chapter for my prestigious "kill X with fire" series? 🤔 🔥 😆 

> After 0.82.0 though. I don't want to touch SDL pain this close to release 😆

Oh yeah, in the 0.83.x timeframe, most definitely 😅

**Comment by GranMinigun** [comment]
The big refactoring commit eliminates the annoying hiccup on toggling fullscreen mode, making the transition feel nice and snappy. Display mode handling (fullscreen, switching resolution) wasn't finished and had reported issues on macOS.

Conflicting functionality is mostly in shader loading code.

Porting to SDL3 won't require any major refactors, unless you want to make it coexist with SDL2. The way the library's APIs are used is mostly the same, just a lot of renamed functions and addition of the properties system (e.g., `SDL_CreateWindowWithProperties()`). Maybe the audio part is different.

**Comment by johnnovak** [comment]
Thanks for stopping by @GranMinigun & for the extra context. Yeah I figured there's not much to change for SDL3 (neither would we benefit much from the extra features because our use of SDL is rather minimal... but we'll need to upgrade eventually, no doubt).

**Comment by weirddan455** [comment]
>  The big refactoring commit eliminates the annoying hiccup on toggling fullscreen mode, making the transition feel nice and snappy. Display mode handling (fullscreen, switching resolution) wasn't finished and had reported issues on macOS.

@johnnovak Didn't you narrow that problem down to shader compilation?  If so I would think shader caching is the important bit here.

**Comment by johnnovak** [comment]
> > The big refactoring commit eliminates the annoying hiccup on toggling fullscreen mode, making the transition feel nice and snappy. Display mode handling (fullscreen, switching resolution) wasn't finished and had reported issues on macOS.
> 
> @johnnovak Didn't you narrow that problem down to shader compilation? If so I would think shader caching is the important bit here.

Yeah, that's one part of it, but probably we're doing some extra unnecessary work on windowed/fullscreen switches, and our favourite Russian got rid of that (and broke the macOS switching somewhat in the simplificstion process).

**Comment by GranMinigun** [comment]
Shader compilation is a different problem, hiccup on fullscreen mode toggle was present even before John introduced automatic CRT shaders. If I remember correctly, that's a regression I introduced at some point.

**Comment by johnnovak** [comment]
The refactors have been addressed in #4449. Combined with my fine-grained config setting update handler rework, now almost all `[sdl]` section settings can be changed at runtime with no flicker in most cases. Fullscreen toggling is a lot smoother too.

Only the shader caching is outstanding.

**Comment by johnnovak** [comment]
Shader caching was reimplemented from scratch in #4551 which concludes this ticket.
